allow "UTC" in string format of timestamp

alexander-tur opened the pull request:
fix problem with statements like:
INSERT INTO tbl (pk, time) VALUES (1, '2016-09-27 16:10:00 UTC');

fixes #20501

**Please replace this line with justification for the backport/\* labels added to this PR**

scylladb-promoter commented:
@scylladb/scylla-maint please review and trigger CI for this pull request
**CI Build:** https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/parambuild/?PR_NUMBER=22426

avikivity commented:
@alexander-tur note that due to our recent [license change](https://www.scylladb.com/2024/12/18/why-were-moving-to-a-source-available-license/) we require a new signed [CLA](https://www.scylladb.com/open-source-nosql-database/contributor-agreement/) to accept contributions.

alexander-tur commented on test/cqlpy/cassandra_tests/functions/operation_fcts_test.py:
```diff
@@ -724,6 +724,17 @@ def testOverflow(cql, test_keyspace):
         assertRows(execute(cql, table, "SELECT a + (int) ?, b + (tinyint) ?, c + (smallint) ? FROM %s", 2147483647, 127, 32767),
                    row(-2147483648, -128, -32768))
 
+def testTinestampWithTimezones(cql, test_keyspace):
```
fixed

alexander-tur commented on types/types.cc:
```diff
@@ -287,19 +287,24 @@ int64_t timestamp_from_string(std::string_view s) {
+        const auto str_begin = tz.find_first_not_of(" ");
+        const auto str_end = tz.find_last_not_of(" ");
+        const auto str_range = str_end - str_begin + 1;
+        auto tz_trim = (str_begin == std::string::npos)? "" : tz.substr(str_begin, str_range);
```
fixed

alexander-tur commented on test/cqlpy/cassandra_tests/functions/operation_fcts_test.py:
```diff
@@ -724,6 +724,17 @@ def testOverflow(cql, test_keyspace):
+    # Test with timestamp type.
+    with create_table(cql, test_keyspace, "(pk int, time timestamp, v int, primary key (pk, time))") as table:
+        execute(cql, table, "INSERT INTO %s (pk, time, v) VALUES (1, '2016-09-25 16:00:00', 1)")
+        execute(cql, table, "INSERT INTO %s (pk, time, v) VALUES (1, '2016-09-27 16:12:00 UTC', 2)")
```
yes we have to trim the string. Otherwise Cassandra test doesn't work.
I've added a test wit space in the end of the string.

scylladb-promoter commented:
@scylladb/scylla-maint please review and trigger CI for this pull request
**CI Build:** https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/parambuild/?PR_NUMBER=22426

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/15081/)
:white_check_mark: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/6386/)
:white_check_mark: - [dtest with tablets](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-tablets/3270/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/14406/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/15081/)
#### Build Details:
- Duration: 5 hr 11 min
- Builder: spider6.cloudius-systems.com

alexander-tur commented:
@nyh 
>     1. Can you please confirm that you tried your new test (the one with the "z" timezone) also on Cassandra (`test/cqpy/run-cassandra`) and it passed?
sorry I missed it and cecked it on Cassandra now. It confused me but Cassandra accept only ' utc' with a single space or 'z' without any spaces as timezone. Do I need to reproduce Cassandra's behavior here? 
> 
>     2. Did you sign the new CLA? I'm sorry for repeatedly asking, I'm still trying to figure out the new process..
Yes, I did.

nyh commented:
> @nyh
> 
> > ```
> > 1. Can you please confirm that you tried your new test (the one with the "z" timezone) also on Cassandra (`test/cqpy/run-cassandra`) and it passed?
> > ```
> 
> sorry I missed it and cecked it on Cassandra now. It confused me but Cassandra accept only ' utc' with a single space or 'z' without any spaces as timezone. Do I need to reproduce Cassandra's behavior here?

I guess it doesn't really matter if Scylla is more **lenient** (excepts everything that Cassandra does, and a bit more), but you needed to make quite a bit of effort for all that "ignore spaces" thing, maybe it's not needed at all? Maybe for UTC you just need to compare to " UTC" (with that extra space in the beginning) and with "z" you just need to compare to "z" (without space), and that will be compatible with Cassandra?

I'll let you decide what you think is better?

By the way, I'm guessing the capital Z also works in Cassandra, can you please check (and if so consider supporting too)?

Finally, could you check if " GMT" also works in Cassandra? For many years, this was the traditional timezone to use before the name UTC somehow became more popular. I've never heard the name "Zulu" outside military contexts ;-)

When I originally opened the issue you were fixing, I wanted to support just "UTC" because that's what was missing to pass Cassandra's unit test. Now that you noticed that "z" should also work, the flood gates have opened - sorry about that ;-)

> > ```
> > 2. Did you sign the new CLA? I'm sorry for repeatedly asking, I'm still trying to figure out the new process..
> > ```
> 
> Yes, I did.

Thanks.

scylladb-promoter commented:
@scylladb/scylla-maint please review and trigger CI for this pull request
**CI Build:** https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/parambuild/?PR_NUMBER=22426

alexander-tur commented:
> > @nyh
> I'll let you decide what you think is better?
Ok 
> 
> By the way, I'm guessing the capital Z also works in Cassandra, can you please check (and if so consider supporting too)?
Yes it works
> 
> Finally, could you check if " GMT" also works in Cassandra? For many years, this was the traditional timezone to use before the name UTC somehow became more popular. I've never heard the name "Zulu" outside military contexts ;-)

Yes. I added "GMT" and fixed tests.

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/15205/)
:white_check_mark: - [dtest with tablets](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-tablets/3368/)
:white_check_mark: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/6483/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/14503/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/15205/)
#### Build Details:
- Duration: 5 hr 8 min
- Builder: spider4.cloudius-systems.com

alexander-tur commented:
@nyh could it be merged now?

denesb commented:
> @nyh could it be merged now?

This PR was queued yesterday, but it didn't promote to master yet, because we have some issues with our CI.
